### PR TITLE
m9 -> m10

### PR DESCRIPTION
**Pull request by hayesgm**
This patch includes the previous m9 changes as a new m10 branch rebased off of m9, which would be force-pushed to develop.

### CODE REVIEW

File: chains/scripts/add_users.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+async function loadem(starting_page = 0) {
```
Comment (jflatow) [inline chains/scripts/add_users.js:1]: Hm what is this file actually?

**Review by jflatow — APPROVED** [review]
I almost forgot how huge this release is. Amazing work 🙌  Let's 🚢 before we change our minds!

File: .github/workflows/test-workflow.yml [hunk 2]
```diff
@@ -40,7 +41,7 @@ jobs:
       - name: Cargo Test
         run: |
           set -eo pipefail
-          cargo test -- -Z unstable-options --format json | tee unit-test-results.json
+          cargo test -- -Z unstable-options --format json --test-threads 1 | tee unit-test-results.json
```
Comment (hayesgm) [inline .github/workflows/test-workflow.yml:44]: Note: we shouldn't _need_ this.

File: .github/workflows/test-workflow.yml [hunk 3]
```diff
@@ -237,14 +238,14 @@ jobs:
 
       - name: Build Gateway
         run: |
-          cargo build
+          cargo build --release --features freeze-time --features runtime-debug
```
Comment (hayesgm) [inline .github/workflows/test-workflow.yml]: Note: we should probably use `--features integration` which properly includes sub-features.

File: README.md [hunk 4]
```diff
@@ -219,4 +219,47 @@ by appending your own. A few useful ones are as follow.
...
+
+Draft a [new release on GitHub](https://github.com/compound-finance/gateway/releases/new).
+Tag it with the appropriate milestone tag.
+Title it in a style similar to other releases, describing its purpose.
+Put any other context or information describing what it does in the description.
+
+Attach the following files to the release, from the repository where you built the release artifacts:
+
+```
+contracts.json
```
Comment (hayesgm) [inline README.md:264]: Note: these are lightly out of date.

File: chains/build_spec.js [hunk 5]
```diff
@@ -97,6 +97,7 @@ async function setAuthorities(chainSpec, chainConfig, opts) {
 async function setStarport(chainSpec, chainConfig, opts) {
   let chainDeployment = await fetchChainDeployment(chainConfig.eth_network);
   let starportAddress = must(chainDeployment.Contracts, 'Starport');
+  // TODO: Set other properties
```
Comment (hayesgm) [inline chains/build_spec.js]: TODO: We should probably remove properties.

File: chains/scripts/add_users.js (hunk 1 above)
Comment (hayesgm) [reply chains/scripts/add_users.js:1]: I think Coburn was using this for... load testing?

File: chains/testnet/chain-spec-raw.json [hunk 6]
```diff
@@ -6,7 +6,12 @@
   "telemetryEndpoints": null,
   "protocolId": "local",
   "properties": {
-    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613"
+    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613",
```
Comment (hayesgm) [inline chains/testnet/chain-spec-raw.json:9]: TODO: Remove properties?

File: chains/testnet/chain-spec.json [hunk 7]
```diff
@@ -6,7 +6,12 @@
   "telemetryEndpoints": null,
   "protocolId": "local",
   "properties": {
-    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613"
+    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613",
```
Comment (hayesgm) [inline chains/testnet/chain-spec.json:9]: TODO: Remove properties?

File: ethereum-client/src/lib.rs [hunk 8]
```diff
@@ -126,96 +199,227 @@ fn send_rpc(
...
-        .map_err(|_| EthereumClientError::JsonParseError)?;
-    let eth_logs = deserialized_body
+    eth_starport_address: &[u8; 20],
+    block_num: EthereumBlockNumber,
+) -> Result<EthereumBlock, EthereumClientError> {
+    let block_str = encode_block_hex(block_num);
+    let block_obj = get_block_object(server, &block_str)?;
+    debug!("eth_starport_address: {:X?}", &eth_starport_address[..]);
+    let get_logs_params = vec![serde_json::json!({
+        "address": format!("0x{}", &eth_starport_address[..].iter().map(|x| format!("{:02x}", x)).collect::<String>()),
```
Comment (hayesgm) [inline ethereum-client/src/lib.rs]: We should remove this hack and just use `::hex::encode` now that we don't do WASM replacements.

File: ethereum-client/src/lib.rs [hunk 9]
```diff
@@ -126,96 +199,227 @@ fn send_rpc(
...
-    parse_u64(Some(
-        deserialized_body
-            .result
-            .ok_or(EthereumClientError::JsonParseError)?,
-    ))
-    .ok_or(EthereumClientError::JsonParseError)
+pub fn get_latest_block_number(server: &str) -> Result<u64, EthereumClientError> {
+    let response_str: String = send_rpc(server, "eth_blockNumber".into(), vec![])?;
+    let response = deserialize_block_number_response(&response_str)?;
+    debug!("eth_blockNumber response: {:?}", response.result.clone());
```
Comment (hayesgm) [inline ethereum-client/src/lib.rs:286]: Not sure this debug statement is actually useful?

File: integration/__tests__/borrow_scen.js [hunk 10]
```diff
@@ -22,13 +21,15 @@ async function lockZRX({ ashley, bert, bat, zrx }) {
 buildScenarios('Borrow Scenarios', borrow_scen_info, { beforeEach: lockZRX }, [
   {
     name: "Borrow BAT and Garner Interest",
-    scenario: async ({ ashley, bert, bat, zrx, chain, starport, log, cash }) => {
+    notes: "This test allows arbitrary time passage, and thus has some estimation",
```
Comment (hayesgm) [inline integration/__tests__/borrow_scen.js:24]: TODO: This should really just use freeze time, now.

File: integration/__tests__/gov_scen.js [hunk 11]
```diff
@@ -21,183 +19,46 @@ buildScenarios('Gov Scenarios', gov_scen_info, [
...
-      let extrinsic = ctx.api().tx.cash.allowNextCodeWithHash(currHash);
+      let extrinsic = api.tx.cash.allowNextCodeWithHash(currHash);
 
-      await starport.executeProposal("Upgrade from m2 to Current [Allow Next Code]", [extrinsic]);
+      await starport.executeProposal("Upgrade from m7 to Current [Allow Next Code]", [extrinsic]);
 
       expect(await chain.nextCodeHash()).toEqual(currHash);
 
-      let event = await chain.setNextCode(await curr.wasm());
+      let event = await chain.setNextCode(await curr.wasm(), curr, false);
```
Comment (hayesgm) [inline integration/__tests__/gov_scen.js:48]: Are these params right?

File: integration/__tests__/liquidate_scen.js [hunk 12]
```diff
@@ -1,46 +1,337 @@
...
+      expect(await bert.chainBalance(comp)).toBeCloseTo(0, 2);
+    }
+  },
+  {
+    name: "Fails when liquidator liquidates too much of borrower's given collateral",
+    notes:
+      `A simple replay of the collateral-for-collateral liquidation scenario, but Ashley
+       doesn't have much of the seized collateral (e.g. she's mostly in USDC, not ETH).
+
+       Note: This test is a counter-factual and should *not* pass`,
```
Comment (hayesgm) [inline integration/__tests__/liquidate_scen.js:247]: This... message should be removed now, right?

File: integration/__tests__/lock_scen.js [hunk 13]
```diff
@@ -109,22 +107,22 @@ buildScenarios('Lock Scenarios', lock_scen_info, [
         ChainAccount: { Eth: ashley.ethAddress().toLowerCase() },
       });
       let data = getEventData(event);
-      expect(data.CashPrincipalAmount).toBeCloseTo(99999996);
+      expect(data.CashPrincipalAmount).toBeCloseTo(99999996, -1); // TODO: Check this better
```
Comment (hayesgm) [inline integration/__tests__/lock_scen.js:110]: TODO: Use freeze-time?

File: integration/__tests__/upgrade_to_m9_m10_scen.js [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+      // start at 0, rotate through 1, actually perform change over on 2
+      await chain.waitUntilSession(2);
+
+      // Try to lock again
+      await ashley.lock(1, usdc);
+      expect(await ashley.chainBalance(usdc)).toEqual(2);
+    }
+  },
+  {
+    only: true, // Currently CI doesnt have native binaries
```
Comment (hayesgm) [inline integration/__tests__/upgrade_to_m9_m10_scen.js]: only -> skip?

File: integration/util/scenario/chain.js [hunk 15]
```diff
@@ -197,16 +227,21 @@ class Chain {
   }
 
   async cullNotices() {
```
Comment (hayesgm) [inline integration/util/scenario/chain.js:229]: TODO: Can we remove this now? We might need to remove earlier integration tests that depended on this function, but that's probably okay?

File: integration/util/scenario/chain.js [hunk 16]
```diff
@@ -254,50 +289,57 @@ class Chain {
...
-    await checkIdx(retries);
+    }, { delay: 1000 });
+  }
+
+  async encodeCall(palletNumber, callNumber, version, argTypes, argValues) {
+    let registry = await version.registry();
+    let argsEncoded = zip(argTypes, argValues).map(([argType, argValue]) => {
+      return registry.createType(argType, argValue).toHex().slice(2);
+    });
+    return `0x04${hexByte(palletNumber)}${hexByte(callNumber)}${argsEncoded.join('')}`;
```
Comment (hayesgm) [inline integration/util/scenario/chain.js]: TODO: This probably should _not_ start with 0x04 (Extrinsic Version)

File: integration/util/scenario/chain_spec.js [hunk 17]
```diff
@@ -48,43 +48,49 @@ async function baseChainSpec(validatorsInfoHash, tokensInfoHash, ctx) {
         full_rate: 2000
       }
     },
-    miner_shares: 1000,
+    miner_shares: 0,
```
Comment (hayesgm) [inline integration/util/scenario/chain_spec.js:51]: TODO: Do we want miner shares for testing? Do we test that, generally?

File: integration/util/scenario/chain_spec.js [hunk 18]
```diff
@@ -48,43 +48,49 @@ async function baseChainSpec(validatorsInfoHash, tokensInfoHash, ctx) {
...
+  let frameSystem = {
+    frameSystem: {
+      code: await genesisVersion.wasm()
+    }
+  };
 
   let reporters = ctx.__reporters();
 
+  let cashGenesis = {};
+  if (genesisVersion.supports('eth-starport-parent-block')) {
```
Comment (hayesgm) [inline integration/util/scenario/chain_spec.js:73]: This is useful.

File: integration/util/scenario/ctx.js [hunk 19]
```diff
@@ -34,10 +37,24 @@ class Ctx {
...
   }
 
-  __initialYieldStart() {
+  __initialYieldStartRaw() {
     return process.env['INITIAL_YIELD_START'] || this.scenInfo['initial_yield_start'] || this.startTime;
   }
 
+  __initialYieldStartMS() {
+    let raw = this.__initialYieldStartRaw();
+    if (raw > 4102444800) { // Jan 1, 2100 as seconds since 1970
```
Comment (hayesgm) [inline integration/util/scenario/ctx.js:46]: Basically, we're just trying to see if the time is so large that it _must_ have been ms and not s

File: integration/util/scenario/validator.js [hunk 20]
```diff
@@ -166,9 +210,11 @@ class Validator {
       '--no-mdns',
       '--node-key',
       this.nodeKey,
-      '-lruntime=debug',
+      ...newCliArgs,
+      '-lruntime=debug,gateway=debug,pallet_cash=debug,ethereum_client=debug',
```
Comment (hayesgm) [inline integration/util/scenario/validator.js:214]: Do we want all of these to be `debug`?

File: scripts/build_release.sh [hunk 21]
```diff
@@ -15,41 +15,53 @@ echo "*** Building release $version ***"
...
-cargo +nightly build --release
+WASM_BUILD_RUSTFLAGS='--cfg feature="testnet"' cargo +nightly build --release --features testnet
 
 bin="./target/release/gateway"
 types="./types.json"
 rpc="./rpc.json"
 wasm="./target/release/wbuild/gateway-runtime/gateway_runtime.compact.wasm"
 
-if [ ! -f "$bin" -o ! -f "$wasm" -o ! -f "$types" -o ! -f "$rpc" -o ! -f "$contracts" ]; then
+if [ ! -f "$bin" -o ! -f "$wasm" -o ! -f "$types" -o ! -f "$rpc" -o ! -f "$contracts" -o ]; then
```
Comment (hayesgm) [inline scripts/build_release.sh]: Dangling `-o`

File: scripts/build_release.sh [hunk 22]
```diff
@@ -15,41 +15,53 @@ echo "*** Building release $version ***"
...
 
 (cd ethereum && yarn install && yarn compile)
 
 contracts="./ethereum/.build/contracts.json"
 
-echo "*** Building release gateway ***"
+echo "*** Building testnet release gateway ***"
 
-cargo +nightly build --release
+WASM_BUILD_RUSTFLAGS='--cfg feature="testnet"' cargo +nightly build --release --features testnet
```
Comment (hayesgm) [inline scripts/build_release.sh]: Is this now testnet? Should it be flagged as such?

File: scripts/build_release.sh [hunk 23]
```diff
@@ -15,41 +15,53 @@ echo "*** Building release $version ***"
...
+echo "$bin_checksum" > "$release_dir/gateway-darwin-arm64.checksum"
+
 cp "$types" "$release_dir/types.json"
 cp "$rpc" "$release_dir/rpc.json"
 cp "$contracts" "$release_dir/contracts.json"
 
 echo "Built release $version"
-echo "  wasm: $release_dir/gateway_runtime.compact.wasm"
-echo "  wasm.checksum: $release_dir/gateway_runtime.checksum"
+echo "  wasm: $release_dir/gateway-testnet.wasm"
```
Comment (hayesgm) [inline scripts/build_release.sh]: This probably _shouldn't_ be tagged test-net now?

File: node/src/chain_spec.rs [hunk 24]
```diff
@@ -99,6 +89,10 @@ fn development_genesis() -> GenesisConfig {
             .duration_since(wasm_timer::UNIX_EPOCH)
             .expect("cannot get system time for genesis")
             .as_millis() as Timestamp,
+        // Starports
+        vec![],
```
Comment (waynenilsen) [inline node/src/chain_spec.rs:93]: This is causing `cargo run -- --dev --tmp --alice` to fail with panic at `pallets/cash/src/lib.rs:713`

```
Thread 'main' panicked at 'Starports must be set in the genesis config',
```

we should probably change something here .... i can look into what we were doing before

**Comment by waynenilsen** [comment]
I also just noticed that in `build_spec.js` the starport address is still being set under the `properties` attribute

```

async function setStarport(chainSpec, chainConfig, opts) {
  let chainDeployment = await fetchChainDeployment(chainConfig.eth_network);
  let starportAddress = must(chainDeployment.Contracts, 'Starport');
  // TODO: Set other properties
  chainSpec.properties.eth_starport_address = starportAddress;
}
```

should be modified

File: chains/build_spec.js (hunk 5 above)
Comment (waynenilsen) [reply chains/build_spec.js]: I'm pretty sure the `properties` key is buried somewhere in substrate but we can remove all of the `properties` that we looked for.

File: .github/workflows/test-workflow.yml (hunk 3 above)
Comment (hayesgm) [reply .github/workflows/test-workflow.yml]: Addressing in my tweaks branch.

File: integration/__tests__/gov_scen.js (hunk 11 above)
Comment (hayesgm) [reply integration/__tests__/gov_scen.js:48]: They are.

File: node/src/chain_spec.rs (hunk 24 above)
Comment (hayesgm) [reply node/src/chain_spec.rs:93]: cool, though maybe it should fail since we need a starport to get going, right?

**Pull request merged**